File: api/scala.with.sbt/octocat-samples/src/main/scala/github/features/Repositories.scala
package github.features

import http.Response

import scala.util.{Failure, Success}
import scala.util.parsing.json.JSON

/** =Repositories features=
  *
  * ==Setup==
  *
  * instantiate the `github.Client` with `Repositories` trait:
  *
  * {{{
  *   val gitHubCli = new github.Client(
  *     "http://github.at.home/api/v3",
  *     sys.env("TOKEN_GITHUB_ENTERPRISE")
  *   ) with Repositories
  *
  * }}}
  */
trait Repositories extends RESTMethods {
  /** Get the list of the repositories for a user
    *
    * @param handle this is the login of the GitHub user
    * @return
    */
  def fetchUserRepositories(handle:String):Either[String, Option[Any]] = {
    getData(s"/users/$handle/repos", generateHeaders.::(new http.Header("Content-Type", "application/json"))) match {
      case Success(resp:Response) =>
        if (http.isOk(resp.code)) Right(JSON.parseFull(resp.data)) else Left(resp.message)
      case Failure(err) =>  Left(err.getMessage)
    }
  }

  /** Get the list of the repositories for an organization
    *
    * @param organization organization name(login)
    * @return
    */
  def fetchOrganizationRepositories(organization:String):Either[String, Option[Any]] = {
    getData(s"/orgs/$organization/repos", generateHeaders.::(new http.Header("Content-Type", "application/json"))) match {
      case Success(resp:Response) =>
        if (http.isOk(resp.code)) Right(JSON.parseFull(resp.data)) else Left(resp.message)
      case Failure(err) =>  Left(err.getMessage)
    }
  }

  /** Create a repository for the authenticated user
    *
    * @param name repository name
    * @param description repository description
    * @param isPrivate set the privacy of the repository
    * @param hasIssues activate or not the issues feature for the repository
    * @return
    */
  def createRepository(name:String, description:String, isPrivate:Boolean, hasIssues:Boolean):Either[String, Option[Any]] = {
    postData(
      "/user/repos",
      generateHeaders.::(new http.Header("Content-Type", "application/json")),
      Map(
        "name" -> name,
        "description" -> description,
        "private" -> isPrivate,
        "has_issues" -> hasIssues,
        "has_wiki" -> true,
        "auto_init" -> true
      )
    ) match {
      case Success(resp:Response) =>
        if (http.isOk(resp.code)) Right(JSON.parseFull(resp.data)) else Left(resp.message)
      case Failure(err) =>  Left(err.getMessage)
    }
  }

  /** Create a repository for an organization
    *
    * @param name repository name
    * @param description repository description
    * @param organization organization name(login)
    * @param isPrivate set the privacy of the repository
    * @param hasIssues activate or not the issues feature for the repository
    * @return
    */
  def createOrganizationRepository(name:String, description:String, organization:String, isPrivate:Boolean, hasIssues:Boolean):Either[String, Option[Any]] = {
    postData(
      s"/orgs/$organization/repos",
      generateHeaders.::(new http.Header("Content-Type", "application/json")),
      Map(
        "name" -> name,
        "description" -> description,
        "private" -> isPrivate,
        "has_issues" -> hasIssues,
        "has_wiki" -> true,
        "auto_init" -> true
      )
    ) match {
      case Success(resp:Response) =>
        if (http.isOk(resp.code)) Right(JSON.parseFull(resp.data)) else Left(resp.message)
      case Failure(err) =>  Left(err.getMessage)
    }
  }
}

File: github-api/javascript/es2015-nodejs/libs/features/repositories.js
/*
# Repositories features

 ## Setup

```javascript
const GitHubClient = require('../libs/GitHubClient.js').GitHubClient;
const repositories = require('../libs/features/repositories');


let githubCli = new GitHubClient({
  baseUri: "http://github.at.home/api/v3",
  token: process.env.TOKEN_GHITHUB_ENTERPRISE
}, repositories); //<-- add repositories features
```
*/

/*
## fetchUserRepositories

- parameter: `handle`
- return: `Promise`

### Description

`fetchUserRepositories` gets the list of the repositories of a user (`handle`)

*/
function fetchUserRepositories({handle}) {
  return this.getData({path:`/users/${handle}/repos`})
    .then(response => {
      return response.data;
    });
}

/*
## fetchOrganizationRepositories

- parameter: `organization` (organization name)
- return: `Promise`

### Description

`fetchOrganizationRepositories` gets the list of the repositories of an organization

*/
function fetchOrganizationRepositories({organization}) {
  return this.getData({path:`/orgs/${organization}/repos`})
    .then(response => {
      return response.data;
    });
}

/*
## createPublicRepository

- parameters: `name, description`
- return: `Promise`

### Description

`createPublicRepository` creates a public repository for the authenticated user

*/
function createPublicRepository({name, description}) {
  return this.postData({path:`/user/repos`, data:{
    name: name,
    description: description,
    private: false,
    has_issue: true,
    has_wiki: true,
    auto_init: true
  }}).then(response => {
    return response.data;
  });
}

/*
## createPrivateRepository

- parameters: `name, description`
- return: `Promise`

### Description

`createPrivateRepository` creates a private repository for the authenticated user

*/
function createPrivateRepository({name, description}) {
  return this.postData({path:`/user/repos`, data:{
    name: name,
    description: description,
    private: true,
    has_issue: true,
    has_wiki: true,
    auto_init: true
  }}).then(response => {
    return response.data;
  });
}

/*
## createPublicOrganizationRepository

- parameters: `name, description, organization`
- return: `Promise`

### Description

`createPublicOrganizationRepository` creates a public repository for an organization

*/
function createPublicOrganizationRepository({name, description, organization}) {
  return this.postData({path:`/orgs/${organization}/repos`, data:{
    name: name,
    description: description,
    private: false,
    has_issue: true,
    has_wiki: true,
    auto_init: true
  }}).then(response => {
    return response.data;
  });
}

/*
## createPrivateOrganizationRepository

- parameters: `name, description, organization`
- return: `Promise`

### Description

`createPrivateOrganizationRepository` creates a private repository for an organization

*/
function createPrivateOrganizationRepository({name, description, organization}) {
  return this.postData({path:`/orgs/${organization}/repos`, data:{
    name: name,
    description: description,
    private: true,
    has_issue: true,
    has_wiki: true,
    auto_init: true
  }}).then(response => {
    return response.data;
  });
}


module.exports = {
  fetchUserRepositories: fetchUserRepositories,
  fetchOrganizationRepositories: fetchOrganizationRepositories,
  createPublicRepository: createPublicRepository,
  createPrivateRepository: createPrivateRepository,
  createPublicOrganizationRepository: createPublicOrganizationRepository,
  createPrivateOrganizationRepository: createPrivateOrganizationRepository
};